deepchem/feat/molecule_featurizers/mat_featurizer.py

```diff
@@ -0,0 +1,103 @@
+import numpy as np
+from rdkit import Chem
+from sklearn.metrics import pairwise_distances
+
+
+class MATFeaturizer(MolecularFeaturizer):
+  """
+  This class is a featurizer for the Molecule Attention Transformer [1]_.
+  The featurizer accepts an RDKit Molecule, and 2 booleans (add_dummy_node and one_hot_formal_charge) as arguments.
+  The returned value is a numpy array which consists of molecular graph descriptions:
+    - Node Features
+    - Adjacency Matrix
+    - Distance Matrix
+
+  References
+  ---------
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer`<https://arxiv.org/abs/2002.08264>`"
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> feat = dc.feat.MATFeaturizer()
+  >>> out = feat.featurize("CCC")
+  """
+
+  def __init__(
+      self,
+      add_dummy_node: bool = True,
+      one_hot_formal_charge: bool = True,
+  ):
+    """
+    Parameters
+    ----------
+    add_dummy_node: bool, default True
+      If True, a dummy node will be added to the molecular graph.
+    one_hot_formal_charge: bool, default True
+      If True, formal charges on atoms are one-hot encoded.
+    """
+
+    self.add_dummy_node = add_dummy_node
+    self.one_hot_formal_charge = one_hot_formal_charge
+
+  def atom_features(self, atom):
+    """
+    Deepchem already contains an atom_features function, however we are defining a new one here due to the need to handle features specific to MAT.
+    Since we need new features like Atom GetNeighbors and IsInRing, and the number of features required for MAT is a fraction of what the Deepchem atom_features function computes, we can speed up computation by defining a custom function.
+    
+    Parameters
+    ----------
+    atom: RDKitAtom
+      RDKit Atom object.
+
+    Returns
+    ----------
+    Atom_features: ndarray
+      Numpy array containing atom features.
+    
+    """
+    attrib = []
+    attrib += one_hot_encode(atom.GetAtomicNum(),
+                             [5, 6, 7, 8, 9, 15, 16, 17, 35, 53, 999])
+    attrib += one_hot_encode(len(atom.GetNeighbors()), [0, 1, 2, 3, 4, 5])
+    attrib += one_hot_encode(atom.GetTotalNumHs(), [0, 1, 2, 3, 4])
+
+    if self.one_hot_formal_charge:
+      attrib += one_hot_encode(atom.GetFormalCharge(), [-1, 0, 1])
+    else:
+      attrib.append(atom.GetFormalCharge())
+
+    attrib.append(atom.IsInRing())
+    attrib.append(atom.GetIsAromatic())
+
+    return np.array(attrib, dtype=np.float32)
+
+  def _featurize(self, mol):
+    """
+    Featurize the molecule.
+
+    Parameters
+    ----------
+    mol: RDKitMol
+      RDKit mol object.
+    
+    Returns
+    -------
+    numpy.ndarray: (node_features, adjacency_matrix, distance_matrix)
+    """
+
+    node_features = np.array(
+        [self.atom_features(atom) for atom in mol.GetAtoms()])
+
+    adjacency_matrix = Chem.rdmolops.GetAdjacencyMatrix(mol)
+
+    distance_matrix = Chem.rdmolops.GetDistanceMatrix(mol)
+
+    adjacency_matrix.resize(node_features.shape)
```

Comment:
Why do we resize `adjacency_matrix` and `distance_matrix` to `node_features.shape`?
I looked at the docs https://numpy.org/doc/stable/reference/generated/numpy.ndarray.resize.html, and I am confused by the example under "Enlarging an array: as above, but missing entries are filled with zeros". I would ideally expect zeros to be added to both rows, but this flattens the matrix, adds zeros, and then reshapes. This is not the desired behavior, I think.

Reply:
Your adjacency and distance matrices have shape (N, N), and the node features matrix has shape (N, F). How would resize work if you have a large molecule where N > F?

Reply:
Yes I had some concerns about this. However, returning the arrays together or as a numpy object doesn't work. I am not sure how else to return them. :(

Reply:
`resize` is the incorrect method. Worst case, you should just concatenate the three matrices to get something of shape (N, 2N+F). Maybe the problem arises because `MATFeaturizer` is treated as a molecule featurizer when it is closer to a graph featurizer.

Reply:
I'm a little confused by the resizing here. Do we expect to add these matrices together in the model? If not, then we can probably just leave them as different sizes right?

Sorry this might be a basic thing I'm missing here :)

Reply:
Me too. Worst case of returning a tuple is causing problems downstream, one can just do concatenation as mentioned above.

Reply:
I think the authors split the dataset into 2 parts post featurization: X had the descriptors and Y had the labels. Thus, concatenation can work I believe. I'll go ahead with this for now then.